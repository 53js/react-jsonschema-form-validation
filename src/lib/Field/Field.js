import classnames from 'classnames';
import memoize from 'memoize-one';
import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';

import Context from '../Form/Context';

class Field extends PureComponent {
	getOnBlurHandler = memoize((touch, name, onBlur) => (event) => {
		touch(name);
		if (onBlur) onBlur(event);
	});

	getOnChangeHandler = memoize((onChange, handleFieldChange) => (event) => {
		if (onChange) {
			// Pass Form.handleFieldChange handler as extra parameter to the onChange handler
			onChange(event, handleFieldChange);
			return;
		}

		handleFieldChange(event);
	});

	render() {
		const {
			children,
			className,
			component: Component,
			onBlur,
			onChange,
			name,
			forwardedRef,
			...props
		} = this.props;

		return (
			<Context.Consumer>
				{form => (
					<Component
						{...props}
						className={
							classnames(
								'Sjf_Field',
								className,
								{
									isInvalid: form.isFieldInvalid(name),
									isSubmitted: form.isSubmitted,
									isTouched: form.isFieldTouched(name),
								},
							)
						}
						name={name}
						onBlur={this.getOnBlurHandler(form.touch, name, onBlur)}
						onChange={this.getOnChangeHandler(onChange, form.handleFieldChange)}
						ref={forwardedRef}
					>
						{children}
					</Component>
				)}
			</Context.Consumer>
		);
	}
}

Field.propTypes = {
	children: PropTypes.node,
	className: PropTypes.string,
	component: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
	forwardedRef: PropTypes.oneOfType([
		PropTypes.shape({}),
		PropTypes.func,
	]),
	onBlur: PropTypes.func,
	onChange: PropTypes.func,
	name: PropTypes.string.isRequired,
};

Field.defaultProps = {
	children: null,
	className: '',
	component: 'input',
	forwardedRef: null,
	onBlur: null,
	onChange: null,
};

// eslint-disable-next-line react/no-multi-comp
export default React.forwardRef((props, ref) => (
	<Field {...props} forwardedRef={ref} />
));
