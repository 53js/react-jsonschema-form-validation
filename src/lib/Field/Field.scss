.Jfv_Field {
	&.isInvalid {
		&.isSubmitted,
		&.isTouched {
			border-color: $color-red;
		}
	}
}
