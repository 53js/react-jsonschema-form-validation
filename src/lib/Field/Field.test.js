import React from 'react';
import { mount } from 'enzyme';

import FormContext from '../Form/Context';
import Field from './Field';

jest.mock('../Form/Context');

it('should match snapshot', () => {
	const field = mount(<Field name="username" />);
	expect(field).toMatchSnapshot();
});

it('should call form.touch when blurred', () => {
	const context = {
		getFieldErrors: jest.fn(() => []),
		handleFieldChange: jest.fn(),
		isFieldInvalid: jest.fn(),
		isFieldTouched: jest.fn(),
		touch: jest.fn(),
	};

	FormContext.Consumer.mockImplementationOnce(props => props.children(context));
	const field = mount(<Field name="username" />);
	field.find('input').simulate('blur');
	expect(context.touch).toHaveBeenCalled();
});

it('should call onBlur handler when blurred', () => {
	const onBlur = jest.fn();
	const field = mount(<Field name="username" onBlur={onBlur} />);
	field.find('input').simulate('blur');
	expect(onBlur).toHaveBeenCalled();
});

it('should add class isSubmitted if form is submitted', () => {
	const context = {
		isFieldInvalid: jest.fn(),
		isFieldTouched: jest.fn(),
	};

	let field = mount(<Field name="username" />);
	expect(field.find('.Jfv_Field').hasClass('isSubmitted')).toBe(false);

	context.isSubmitted = true;
	FormContext.Consumer.mockImplementationOnce(props => props.children(context));
	field = mount(<Field name="username" />);
	expect(field.find('.Jfv_Field').hasClass('isSubmitted')).toBe(true);
});

it('should add class isTouched if field is touched', () => {
	const context = {
		isFieldInvalid: jest.fn(),
		isFieldTouched: jest.fn(),
	};

	let field = mount(<Field name="username" />);
	expect(field.find('.Jfv_Field').hasClass('isTouched')).toBe(false);

	context.isFieldTouched.mockImplementation(() => true);
	FormContext.Consumer.mockImplementationOnce(props => props.children(context));
	field = mount(<Field name="username" />);
	expect(field.find('.Jfv_Field').hasClass('isTouched')).toBe(true);
});

it('should add class isInvalid if field is invalid', () => {
	const context = {
		isFieldInvalid: jest.fn(),
		isFieldTouched: jest.fn(),
	};

	let field = mount(<Field name="username" />);
	expect(field.find('.Jfv_Field').hasClass('isInvalid')).toBe(false);

	context.isFieldInvalid.mockImplementation(() => true);
	FormContext.Consumer.mockImplementationOnce(props => props.children(context));
	field = mount(<Field name="username" />);
	expect(field.find('.Jfv_Field').hasClass('isInvalid')).toBe(true);
});

it('Default component input can be changed', () => {
	const Component = () => 'component';
	const field = mount(<Field component={Component} name="username" />);
	expect(field.exists(Component)).toBe(true);
});

it('should call form.handleFieldChange() when field value changes', () => {
	const context = {
		handleFieldChange: jest.fn(),
		isFieldInvalid: jest.fn(),
		isFieldTouched: jest.fn(),
	};
	FormContext.Consumer.mockImplementationOnce(props => props.children(context));

	const field = mount(<Field type="text" name="username" />);
	const input = field.find('input');
	input.getDOMNode().value = 'newvalue';
	input.simulate('change');
	expect(context.handleFieldChange).toHaveBeenCalledWith(
		expect.objectContaining({
			target: expect.objectContaining({
				name: 'username',
				value: 'newvalue',
			}),
		}),
	);
});

it('should call onChange handler passed as prop when field value changes', () => {
	const context = {
		handleFieldChange: jest.fn(),
		isFieldInvalid: jest.fn(),
		isFieldTouched: jest.fn(),
	};
	FormContext.Consumer.mockImplementationOnce(props => props.children(context));
	const handleChange = jest.fn();
	const field = mount(<Field type="text" onChange={handleChange} name="username" />);
	const input = field.find('input');
	input.getDOMNode().value = 'newvalue';
	input.simulate('change');
	expect(handleChange).toHaveBeenCalledWith(
		expect.objectContaining({
			target: expect.objectContaining({
				name: 'username',
				value: 'newvalue',
			}),
		}),
		context.handleFieldChange,
	);
	expect(context.handleFieldChange).not.toHaveBeenCalled();
});
