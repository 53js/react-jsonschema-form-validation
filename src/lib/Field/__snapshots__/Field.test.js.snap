// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<ForwardRef
  name="username"
>
  <Field
    className=""
    component="input"
    forwardedRef={null}
    name="username"
    onBlur={null}
    onChange={null}
  >
    <mockConstructor>
      <input
        className="Sjf_Field"
        name="username"
        onBlur={[Function]}
        onChange={[Function]}
      />
    </mockConstructor>
  </Field>
</ForwardRef>
`;
